File: 2018_prelim/tasks/snail/solutions/official_clarence/snail.cpp
#include <cstdio>
#include <algorithm>
#include <assert.h>
#define INF 1012345678012345678LL
#define BIG 1000000000000LL
using namespace std;
typedef long long ll;
// height, number of phases, phase
ll H;
int N;
ll P[10010];
ll simulate[20010];
ll increment = 0; // increment every cycle
bool printed = 0;
int main(void) {
    scanf("%lld%d", &H, &N);
    assert(1<=H && H<=BIG);
    assert(1<=N && N<=10000);
    for (int i=0;i<N;i++) {
        scanf("%lld", &P[i]);
        assert(-BIG<=P[i] && P[i]<=BIG);
    }
    // simulate for 2 days
    simulate[0] = 0; // height at start of each phase
    for (int i=0;i<2*N;i++) { // people may forget this
        simulate[i+1] = max(simulate[i]+P[i%N], 0LL); // people may leave out max
        if (simulate[i+1] >= H) { // >= is intended
            printf("%d %d\n", i/N, i%N);
            printed = 1;
            break;
        }
    }
    if (!printed) {
        for (int i=0;i<N;i++) {
            increment += P[i]; // should not overflow, but can underflow
            // oh no what if it underflows?
            if (increment <= -H) { // if it ever goes back above 0, then it must be possible within first 2N phases
                printf("-1 -1\n");
                printed = true;
                break;
            }
        }
    }
    if (!printed) {
        // check if it goes negative anywhere
        if (increment <= 0) {
            // not possible anymore
            printf("-1 -1\n");
            printed = true;
        } else {
            ll bestDays = INF;
            int phase = 0;
            for (int i=0;i<N;i++) {
                ll days = 1+(H-simulate[N+i+1]+increment-1)/increment;
                if (bestDays > days) {
                    bestDays = days;
                    phase = i;
                }
            }
            printf("%lld %d\n", bestDays, phase);
            printed = true;
        }
    }
    return 0;
}

File: 2018_prelim/tasks/snail/solutions/official_clarence/snail.py
raw_input = input # python 2?
from builtins import input

INF = 1012345678012345678
BIG = 1000000000000
H, N = map(int, input().split())
assert(1<=H and H<=BIG)
assert(1<=N and N<=10000)
printed = False
P = []
simulate = []
# input data into P
P = list(map(int, input().split()))
# simulate for 2 days
simulate.append(0) # height at start of each phase
for i in range(2*N): # people may forget this
    simulate.append(max(simulate[i]+P[i%N], 0)) # people may leave out max
    if simulate[i+1] >= H: # >= is intended
        print("%d %d\n" % (i/N, i%N))
        printed = True
        break

increment = 0
if not printed:
    for i in range(N):
        increment += P[i]
    if increment <= 0:
        # not possible anymore
        print("-1 -1\n")
        printed = True

if not printed:
    bestDays = INF
    phase = 0
    for i in range(N):
        days = 1+(H-simulate[N+i+1]+increment-1)//increment;
        if bestDays > days:
            bestDays = days
            phase = i
    print("%d %d\n" % (bestDays, phase))
    printed = True
